fiksi/src/analyze/graph/recursive_assembly.rs

```diff
@@ -442,103 +442,162 @@ pub(crate) fn decompose<const D: i16>(
+
+    // Seed BFS with all singleton (size 1) subgraphs.
+    for &vertex in vertices.iter() {
+        let mut subgraph = HashSet::with_capacity(4);
+        subgraph.insert(vertex);
+
+        let mut adjacent_vertices = HashSet::with_capacity(8);
+        extend_adjacent_vertices(
+            &mut adjacent_vertices,
+            graph,
+            available_edges,
+            vertex,
+            &subgraph,
+            vertices,
+        );
+
+        queue.push_back(SubgraphState {
+            subgraph,
+            dof: graph.elements[vertex.id as usize].dof,
+            adjacent_vertices,
+        });
+    }
+
+    while let Some(SubgraphState {
+        subgraph: sub,
+        dof,
+        adjacent_vertices,
+    }) = queue.pop_front()
+    {
+        for &vertex in adjacent_vertices.iter() {
+            // Build next state.
+            let mut next_subgraph = sub.clone();
+            next_subgraph.insert(vertex);
+            debug_assert!(
+                next_subgraph.len() >= 2,
+                "We should never be considering trivial subgraphs of 1 element."
+            );
+
+            let valency = additional_valency(graph, available_edges, &next_subgraph, vertex);
+            let next_dof = dof + graph.elements[vertex.id as usize].dof - valency;
+
+            if !blocked_subgraphs.contains(&next_subgraph) && next_dof > k {
+                return next_subgraph;
             }
 
-            if recurse::<D>(
+            // Find the adjacent vertices of `next_subgraph` by removing the vertex we just added,
+            // and adding the vertices adjacent to the vertex we just added.
+            let mut next_adjacent_vertices = adjacent_vertices.clone();
+            next_adjacent_vertices.remove(&vertex);
+            extend_adjacent_vertices(
+                &mut next_adjacent_vertices,
                 graph,
-                blocked_subgraphs,
                 available_edges,
-                available_vertices,
-                subgraph,
-                dof_vertices,
-                valency_edges,
-            ) {
-                return true;
-            }
-
-            subgraph.remove(&vertex);
-            available_vertices.insert(vertex);
+                vertex,
+                &next_subgraph,
+                vertices,
+            );
+
+            queue.push_back(SubgraphState {
+                subgraph: next_subgraph,
+                dof: next_dof,
+                adjacent_vertices: next_adjacent_vertices,
+            });
         }
-        false
     }
 
-    if recurse::<D>(
-        graph,
-        blocked_subgraphs,
-        available_edges,
-        &mut available_vertices,
-        &mut subgraph,
-        0,
-        0,
-    ) {
-        subgraph
-    } else {
-        // No subgraph found. This means all subgraphs are underconstrained.
-        HashSet::new()
-    }
+    // No subgraph found. This means all subgraphs are underconstrained.
+    HashSet::new()
```

Comment:
Feel like a better return type is needed here ... should this be an `Option`?

Reply:
Indeed...